runtime/src/main/java/io/micronaut/runtime/http/scope/RequestCustomScope.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2017-2019 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.runtime.http.scope;
+
+import io.micronaut.context.BeanContext;
+import io.micronaut.context.BeanResolutionContext;
+import io.micronaut.context.LifeCycle;
+import io.micronaut.context.event.ApplicationEventListener;
+import io.micronaut.context.exceptions.NoSuchBeanException;
+import io.micronaut.context.scope.CustomScope;
+import io.micronaut.core.convert.value.MutableConvertibleValues;
+import io.micronaut.core.util.ArgumentUtils;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.context.ServerRequestContext;
+import io.micronaut.http.context.event.HttpRequestTerminatedEvent;
+import io.micronaut.inject.BeanDefinition;
+import io.micronaut.inject.BeanIdentifier;
+import io.micronaut.inject.DisposableBeanDefinition;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A {@link CustomScope} that creates a new bean for every HTTP request.
+ *
+ * @author James Kleeh
+ * @author Marcel Overdijk
+ * @since 1.2.0
+ */
+@Singleton
+class RequestCustomScope implements CustomScope<RequestScope>, LifeCycle<RequestCustomScope>, ApplicationEventListener<HttpRequestTerminatedEvent> {
```

Comment:
Is the class in general supposed to be thread safe? If so, I think `getRequestScopedBeans` should likely be synchronized since it is not atomic